### PR TITLE
DRILL-4131: Move RPC allocators under Drill's root allocator & accounting

### PR DESCRIPTION
**Pull request by jacques-n**
- Allow settings to be set to ensure RPC reservation and maximums (currently unset by default). Defaults set in drill-module.conf
- Add new metrics to report RPC layer memory consumption.
- Check for memory leaks from RPC layer at shutdown.

### CODE REVIEW

File: exec/java-exec/src/main/java/org/apache/drill/exec/service/ServiceEngine.java [hunk 1]
```diff
@@ -56,22 +60,83 @@
...
+  private final BufferAllocator dataAllocator;
+
 
   public ServiceEngine(ControlMessageHandler controlMessageHandler, UserWorker userWorker, BootStrapContext context,
       WorkEventBus workBus, WorkerBee bee, boolean allowPortHunting) throws DrillbitStartupException {
+    userAllocator = newAllocator(context, "rpc:user", "drill.exec.rpc.user.server.memory.reservation",
+        "drill.exec.rpc.user.server.memory.maximum");
+    controlAllocator = newAllocator(context, "rpc:bit-control",
+        "drill.exec.rpc.bit.server.memory.control.reservation", "drill.exec.rpc.bit.server.memory.control.maximum");
+    dataAllocator = newAllocator(context, "rpc:bit-control",
```
Comment (adeneche) [inline exec/java-exec/src/main/java/org/apache/drill/exec/service/ServiceEngine.java]: should it be named `rpc:bit-data` ?

File: exec/memory/base/src/main/java/io/netty/buffer/PooledByteBufAllocatorL.java [hunk 2]
```diff
@@ -59,7 +59,7 @@ public PooledByteBufAllocatorL(MetricRegistry registry) {
 
   public UnsafeDirectLittleEndian allocate(int size) {
     try {
-      return allocator.directBuffer(size, size);
+      return allocator.directBuffer(size, Integer.MAX_VALUE);
```
Comment (adeneche) [inline exec/memory/base/src/main/java/io/netty/buffer/PooledByteBufAllocatorL.java]: Out of curiosity, why this change ?

**Comment by adeneche** [comment]
apart from a small code change, LGTM. 
Let's run a couple of tests (performance and concurrency) before merging this PR.

+1

File: exec/java-exec/src/main/java/org/apache/drill/exec/service/ServiceEngine.java (hunk 1 above)
Comment (jacques-n) [inline exec/java-exec/src/main/java/org/apache/drill/exec/service/ServiceEngine.java]: nice catch

File: exec/memory/base/src/main/java/io/netty/buffer/PooledByteBufAllocatorL.java (hunk 2 above)
Comment (jacques-n) [inline exec/memory/base/src/main/java/io/netty/buffer/PooledByteBufAllocatorL.java]: The RPC uses ensureWritable to expand buffers. Buffers are expandable to maxCapacity. We were coding maxCapacity == capacity which meant that buffers weren't expandable and the RPC layer would error out. This fixes this issue.

**Comment by jacques-n** [comment]
One quick note, I still am receiving failures on a small set of unit tests because of metric registration conflicts so I'll be fixing those. (Happens only in tests where a single JVM contains multiple drillbits since the registry is currently static.) Also, I'm concerned that some portion of allocation is not getting tracked since the tests that Vicky just posted on DRILL-4266 don't show a zero peak allocation for the RPC data layer (which doesn't make any sense unless all the queries were single fragmented).

File: exec/java-exec/src/main/java/org/apache/drill/exec/service/ServiceEngine.java (hunk 1 above)
Comment (sudheeshkatkam) [inline exec/java-exec/src/main/java/org/apache/drill/exec/service/ServiceEngine.java]: Can you please move the reservation and maximum string constants to ExecConstants?

Comment (jacques-n) [inline exec/java-exec/src/main/java/org/apache/drill/exec/service/ServiceEngine.java]: Someone argued recently that it doesn't make sense to put everything there and I am in agreement. I've started adopting keeping strings where they are relevant.
